### ISSUE TITLE
Opening Modal Crashes App: UIApplicationInvalidInterfaceOrientation

### ISSUE THREAD

**Issue by evansobkowicz** [clearsightstudio/ProMotion#607]
My app crashes any time I call ProMotion's open_modal. 

I have these lines of code in my app, but I've done some googling I can't figure out if this is the problem or not. Any help would be great!

```ruby
  def shouldAutorotate
    true
  end

  def supportedInterfaceOrientations
    if ipad?
      UIInterfaceOrientationMaskLandscapeLeft | UIInterfaceOrientationMaskPortrait | UIInterfaceOrientationMaskLandscapeRight
    else
      UIInterfaceOrientationMaskPortrait
    end
  end

  def preferredInterfaceOrientationForPresentation
    ipad? ? UIInterfaceOrientationMaskLandscapeLeft : UIInterfaceOrientationMaskPortrait
  end
```

```
*** Terminating app due to uncaught exception 'UIApplicationInvalidInterfaceOrientation', reason: 'preferredInterfaceOrientationForPresentation must return a supported interface orientation!'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000103a1bf35 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000100f20bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x0000000103a1be6d +[NSException raise:format:] + 205
	3   UIKit                               0x00000001027624d7 -[UIViewController _preferredInterfaceOrientationForPresentationInWindow:fromInterfaceOrientation:] + 487
	4   UIKit                               0x0000000102cefc72 -[_UIFullscreenPresentationController _adjustOrientationIfNecessaryInWindow:forViewController:preservingViewController:] + 312
	5   UIKit                               0x000000010272b3d2 -[UIPresentationController _presentWithAnimationController:interactionController:target:didEndSelector:] + 667
	6   UIKit                               0x0000000102759dee -[UIViewController _presentViewController:modalSourceViewController:presentationController:animationController:interactionController:completion:] + 732
	7   UIKit                               0x000000010275ae99 -[UIViewController _presentViewController:withAnimationController:completion:] + 2858
	8   UIKit                               0x000000010275cd81 __62-[UIViewController presentViewController:animated:completion:]_block_invoke + 132
	9   UIKit                               0x000000010275cca5 -[UIViewController presentViewController:animated:completion:] + 229
	10  The Ithacan                         0x00000001001b8086 __unnamed_352 + 214
	11  The Ithacan                         0x000000010062f428 rb_vm_dispatch + 6280
	12  The Ithacan                         0x0000000100111cb3 vm_dispatch + 1347
	13  The Ithacan                         0x00000001003741cc rb_scope__present_modal_view_controller:__ + 380
	14  The Ithacan                         0x000000010062f85c rb_vm_dispatch + 7356
	15  The Ithacan                         0x0000000100111cb3 vm_dispatch + 1347
	16  The Ithacan                         0x00000001003719a0 rb_scope__open_screen:__ + 1136
	17  The Ithacan                         0x000000010062f85c rb_vm_dispatch + 7356
	18  The Ithacan                         0x0000000100111cb3 vm_dispatch + 1347
	19  The Ithacan                         0x0000000100372300 rb_scope__open_modal:__ + 272
	20  The Ithacan                         0x000000010062f85c rb_vm_dispatch + 7356
	21  The Ithacan                         0x0000000100111cb3 vm_dispatch + 1347
	22  The Ithacan                         0x00000001004e0f1e rb_scope__open_website:__ + 798
	23  The Ithacan                         0x000000010062f85c rb_vm_dispatch + 7356
	24  The Ithacan                         0x0000000100626195 send_internal + 741
	25  The Ithacan                         0x000000010062f292 rb_vm_dispatch + 5874
	26  The Ithacan                         0x0000000100111cb3 vm_dispatch + 1347
	27  The Ithacan                         0x00000001003ba1b1 rb_scope__trigger_action:__ + 1009
	28  The Ithacan                         0x000000010062f85c rb_vm_dispatch + 7356
	29  The Ithacan                         0x0000000100111cb3 vm_dispatch + 1347
	30  The Ithacan                         0x00000001003bdee8 rb_scope__tableView:didSelectRowAtIndexPath:__ + 600
	31  The Ithacan                         0x00000001003bdf6d __unnamed_167 + 61
	32  CoreFoundation                      0x0000000103913c8c __invoking___ + 140
	33  CoreFoundation                      0x0000000103913ae2 -[NSInvocation invoke] + 290
	34  CoreFoundation                      0x00000001039a2ed6 -[NSInvocation invokeWithTarget:] + 54
	35  The Ithacan                         0x000000010009aebd -[PXProxy forwardInvocation:] + 495
	36  CoreFoundation                      0x000000010397b08f ___forwarding___ + 495
	37  CoreFoundation                      0x000000010397ae18 _CF_forwarding_prep_0 + 120
	38  UIKit                               0x0000000102711393 -[UITableView _selectRowAtIndexPath:animated:scrollPosition:notifyDelegate:] + 1293
	39  UIKit                               0x00000001027114d4 -[UITableView _userSelectRowAtPendingSelectionIndexPath:] + 219
	40  UIKit                               0x000000010264c331 _applyBlockToCFArrayCopiedToStack + 314
	41  UIKit                               0x000000010264c1ab _afterCACommitHandler + 516
	42  CoreFoundation                      0x0000000103950dc7 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	43  CoreFoundation                      0x0000000103950d20 __CFRunLoopDoObservers + 368
	44  CoreFoundation                      0x0000000103946b53 __CFRunLoopRun + 1123
	45  CoreFoundation                      0x0000000103946486 CFRunLoopRunSpecific + 470
	46  GraphicsServices                    0x000000010778f9f0 GSEventRunModal + 161
	47  UIKit                               0x0000000102629420 UIApplicationMain + 1282
	48  The Ithacan                         0x000000010011ab25 main + 117
	49  libdyld.dylib                       0x00000001060b8145 start + 1
	50  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by markrickert** [comment]
Looks like your app's info plist file doesn't have the correct array of supported interface orientations. 

Make sure you have this line in your `Rakefile`:

```ruby
  # leave out :portrait_upside_down since you don't seem to use it anywhere
  app.interface_orientations = [:portrait, :landscape_left, :landscape_right]
```

Hopefully that will fix your issue.

**Comment by evansobkowicz** [comment]
Thanks for the quick reply! I had that line in the past, and I just put it back. It didn't solve the problem though...

```
2015-01-24 13:17:03.790 The Ithacan[17315:1179353] *** Terminating app due to uncaught exception 'UIApplicationInvalidInterfaceOrientation', reason: 'preferredInterfaceOrientationForPresentation must return a supported interface orientation!'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000103a1bf35 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000100f20bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x0000000103a1be6d +[NSException raise:format:] + 205
	3   UIKit                               0x00000001027624d7 -[UIViewController _preferredInterfaceOrientationForPresentationInWindow:fromInterfaceOrientation:] + 487
	4   UIKit                               0x0000000102cefc72 -[_UIFullscreenPresentationController _adjustOrientationIfNecessaryInWindow:forViewController:preservingViewController:] + 312
	5   UIKit                               0x000000010272b3d2 -[UIPresentationController _presentWithAnimationController:interactionController:target:didEndSelector:] + 667
	6   UIKit                               0x0000000102759dee -[UIViewController _presentViewController:modalSourceViewController:presentationController:animationController:interactionController:completion:] + 732
	7   UIKit                               0x000000010275ae99 -[UIViewController _presentViewController:withAnimationController:completion:] + 2858
	8   UIKit                               0x000000010275cd81 __62-[UIViewController presentViewController:animated:completion:]_block_invoke + 132
	9   UIKit                               0x000000010275cca5 -[UIViewController presentViewController:animated:completion:] + 229
	10  The Ithacan                         0x00000001001b8066 __unnamed_352 + 214
	11  The Ithacan                         0x000000010062f438 rb_vm_dispatch + 6280
	12  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	13  The Ithacan                         0x00000001003741ac rb_scope__present_modal_view_controller:__ + 380
	14  The Ithacan                         0x000000010062f86c rb_vm_dispatch + 7356
	15  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	16  The Ithacan                         0x0000000100371980 rb_scope__open_screen:__ + 1136
	17  The Ithacan                         0x000000010062f86c rb_vm_dispatch + 7356
	18  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	19  The Ithacan                         0x00000001003722e0 rb_scope__open_modal:__ + 272
	20  The Ithacan                         0x000000010062f86c rb_vm_dispatch + 7356
	21  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	22  The Ithacan                         0x000000010050ac4c rb_scope__on_submit:__ + 732
	23  The Ithacan                         0x000000010062f86c rb_vm_dispatch + 7356
	24  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	25  The Ithacan                         0x00000001004d882c rb_scope__new:__block__ + 172
	26  The Ithacan                         0x0000000100630b30 _ZL13vm_block_evalP7RoxorVMP11rb_vm_blockP13objc_selectormiPKm + 1024
	27  The Ithacan                         0x000000010062f2a2 rb_vm_dispatch + 5874
	28  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	29  The Ithacan                         0x0000000100403dfd rb_scope__submit__ + 621
	30  The Ithacan                         0x000000010062f86c rb_vm_dispatch + 7356
	31  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	32  The Ithacan                         0x00000001004aaa80 rb_scope__on_select:__ + 96
	33  The Ithacan                         0x000000010062f86c rb_vm_dispatch + 7356
	34  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	35  The Ithacan                         0x00000001003e9e3b rb_scope___on_select:__ + 587
	36  The Ithacan                         0x000000010062f86c rb_vm_dispatch + 7356
	37  The Ithacan                         0x0000000100111c93 vm_dispatch + 1347
	38  The Ithacan                         0x000000010040e2f4 rb_scope__tableView:didSelectRowAtIndexPath:__ + 404
	39  The Ithacan                         0x000000010040e33d __unnamed_101 + 61
	40  CoreFoundation                      0x0000000103913c8c __invoking___ + 140
	41  CoreFoundation                      0x0000000103913ae2 -[NSInvocation invoke] + 290
	42  CoreFoundation                      0x00000001039a2ed6 -[NSInvocation invokeWithTarget:] + 54
	43  The Ithacan                         0x000000010009ae9d -[PXProxy forwardInvocation:] + 495
	44  CoreFoundation                      0x000000010397b08f ___forwarding___ + 495
	45  CoreFoundation                      0x000000010397ae18 _CF_forwarding_prep_0 + 120
	46  UIKit                               0x0000000102711393 -[UITableView _selectRowAtIndexPath:animated:scrollPosition:notifyDelegate:] + 1293
	47  UIKit                               0x00000001027114d4 -[UITableView _userSelectRowAtPendingSelectionIndexPath:] + 219
	48  UIKit                               0x000000010264c331 _applyBlockToCFArrayCopiedToStack + 314
	49  UIKit                               0x000000010264c1ab _afterCACommitHandler + 516
	50  CoreFoundation                      0x0000000103950dc7 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	51  CoreFoundation                      0x0000000103950d20 __CFRunLoopDoObservers + 368
	52  CoreFoundation                      0x0000000103946b53 __CFRunLoopRun + 1123
	53  CoreFoundation                      0x0000000103946486 CFRunLoopRunSpecific + 470
	54  GraphicsServices                    0x000000010778f9f0 GSEventRunModal + 161
	55  UIKit                               0x0000000102629420 UIApplicationMain + 1282
	56  The Ithacan                         0x000000010011ab05 main + 117
	57  libdyld.dylib                       0x00000001060b8145 start + 1
	58  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by jamonholmgren** [comment]
Could you create a minimal replication app using `promotion new <name>`? I'll pull it down and see what's going on.

**Comment by evansobkowicz** [comment]
I think I may have found the issue. I had the interface orientation methods on most of my screens, including the modal screens. Removing those seems to have fixed my issue.

I did make a demo app, and you should be able to comment/uncomment the include line on the HelpScreen to see the bug. https://github.com/evansobkowicz/RubyMotionModalBug

I haven't been able to find a way to allow only the portrait orientation on an iPhone, but allow all orientations on an iPad. Any advice with that?

**Comment by jamonholmgren** [comment]
Cool, thanks @evansobkowicz.

@silasjmatson or @kevinvangelder, could you look into this?

**Comment by kevinvangelder** [comment]
@evansobkowicz What iOS version are you running your simulator on? The iOS docs say that `UIInterfaceOrientationMask`s should be available in iOS 6+, but by using `UIInterfaceOrientation` (no `Mask`, iOS 8+) instead I was able to get the modal to open successfully, so I'm guessing their docs are out of date.

```ruby
  def preferredInterfaceOrientationForPresentation
    ipad? ? UIInterfaceOrientationLandscapeLeft : UIInterfaceOrientationPortrait
  end
```

It looks like the underlying reason for this is that `UIInterfaceOrientationMaskPortrait` is returning `2` instead of `1`, which apparently is not a supported interface orientation.

Finally, I also verified that since you are handling orientation manually, setting `app.interface_orientations` in your Rakefile is unnecessary.
